opm/simulators/linalg/bda/Reorder.cpp

```diff
@@ -0,0 +1,372 @@
+
+template <unsigned int block_size>
+void blocked_reorder_matrix_by_pattern(BlockedMatrix *mat, int *toOrder, int *fromOrder, BlockedMatrix *rMat) {
+    const unsigned int bs = block_size;
+    int rIndex = 0;
+    int i, k;
+    unsigned int j;
+
+    rMat->rowPointers[0] = 0;
+    for (i = 0; i < mat->Nb; i++) {
+        int thisRow = fromOrder[i];
+        // put thisRow from the old matrix into row i of the new matrix
+        rMat->rowPointers[i + 1] = rMat->rowPointers[i] + mat->rowPointers[thisRow + 1] - mat->rowPointers[thisRow];
+        for (k = mat->rowPointers[thisRow]; k < mat->rowPointers[thisRow + 1]; k++) {
+            for (j = 0; j < bs * bs; j++){
+                rMat->nnzValues[rIndex * bs * bs + j] = mat->nnzValues[k * bs * bs + j];
+            }
+            rMat->colIndices[rIndex] = mat->colIndices[k];
+            rIndex++;
+        }
+    }
+    // re-assign column indices according to the new positions of the nodes referenced by the column indices
+    for (i = 0; i < mat->nnzbs; i++) {
+        rMat->colIndices[i] = toOrder[rMat->colIndices[i]];
+    }
+    // re-sort the column indices of every row.
+    for (i = 0; i < mat->Nb; i++) {
+        sortBlockedRow<bs>(rMat->colIndices, rMat->nnzValues, rMat->rowPointers[i], rMat->rowPointers[i + 1] - 1);
+    }
+}
+
+/* Reorder a matrix according to the colors that every node of the matrix has received*/
+
+void colorsToReordering(int Nb, std::vector<int>& colors, int *toOrder, int *fromOrder, int *iters) {
+    int reordered = 0;
+    int i, c;
+    for (i = 0; i < MAX_COLORS; i++) {
+        iters[i] = 0;
+    }
+
+    // Find reordering patterns
+    for (c = 0; c < MAX_COLORS; c++) {
+        for (i = 0; i < Nb; i++) {
+            if (colors[i] == c) {
+                iters[c]++;
+                toOrder[i] = reordered;
+
+                fromOrder[reordered] = i;
+                reordered++;
+            }
+        }
+    }
+}
+
+// Reorder a matrix according to a reordering pattern
+
+template <unsigned int block_size>
+void blocked_reorder_vector_by_pattern(int Nb, double *vector, int *fromOrder, double *rVector) {
+    for (int i = 0; i < Nb; i++) {
+        for (unsigned int j = 0; j < block_size; j++) {
+            rVector[block_size * i + j] = vector[block_size * fromOrder[i] + j];
+        }
+    }
+}
+
+
+/* Check is operations on a node in the matrix can be started
+ * A node can only be started if all nodes that it depends on during sequential execution have already completed.*/
+
+bool canBeStarted(int rowIndex, int *rowPointers, int *colIndices, std::vector<bool>& doneRows) {
+    bool canStart = !doneRows[rowIndex];
+    int i, thisDependency;
+    if (canStart) {
+        for (i = rowPointers[rowIndex]; i < rowPointers[rowIndex + 1]; i++) {
+            thisDependency = colIndices[i];
+            // Only dependencies on rows that should execute before the current one are relevant
```

Comment:
Thank you for pointing this out: this is an error we never caught because of inadequate testing with non-square matrices. The confusing naming of the CSC arrays has probably contributed to this error existing, so in upcoming changes to this PR, we changed the names to CSCRowIndices and CSCColPointers. The canBeStarted function will then be called with: `canBeStarted(thisRow, CSCColPointers, CSCRowIndices, doneRows)`